gadget/mkfs_test.go

```diff
@@ -0,0 +1,194 @@
+	m.cleanup = append(m.cleanup, func() { cmdFakeroot.Restore() })
+
+	cmdMkfsExt4 := testutil.MockCommand(c, "mkfs.ext4", "echo 'override in test' ; exit 1")
+	m.cleanup = append(m.cleanup, func() { cmdMkfsExt4.Restore() })
+
+	cmdMkfsVfat := testutil.MockCommand(c, "mkfs.vfat", "echo 'override in test'; exit 1")
+	m.cleanup = append(m.cleanup, func() { cmdMkfsVfat.Restore() })
+
+	cmdMcopy := testutil.MockCommand(c, "mcopy", "echo 'override in test'; exit 1")
+	m.cleanup = append(m.cleanup, func() { cmdMcopy.Restore() })
+}
+
+func (m *mkfsSuite) TearDownTest(c *C) {
+	for _, r := range m.cleanup {
+		r()
+	}
+}
+
+func (m *mkfsSuite) TestMkfsExt4Happy(c *C) {
+	cmd := testutil.MockCommand(c, "fakeroot", "")
+	defer cmd.Restore()
+
+	err := gadget.MkfsExt4("foo.img", "my-label", "contents")
+	c.Assert(err, IsNil)
+	c.Check(cmd.Calls(), DeepEquals, [][]string{
+		{
+			"fakeroot",
+			"mkfs.ext4",
+			"-T", "default",
+			"-O", "-metadata_csum",
+			"-O", "uninit_bg",
+			"-d", "contents",
+			"-L", "my-label",
+			"foo.img",
+		},
+	})
+
+	cmd.ForgetCalls()
+
+	// empty label
+	err = gadget.MkfsExt4("foo.img", "", "contents")
+	c.Assert(err, IsNil)
+	c.Check(cmd.Calls(), DeepEquals, [][]string{
+		{
+			"fakeroot",
+			"mkfs.ext4",
+			"-T", "default",
+			"-O", "-metadata_csum",
+			"-O", "uninit_bg",
+			"-d", "contents",
+			"foo.img",
+		},
+	})
+}
+
+func (m *mkfsSuite) TestMkfsExt4Error(c *C) {
+	cmd := testutil.MockCommand(c, "fakeroot", "echo 'command failed'; exit 1")
+	defer cmd.Restore()
+
+	err := gadget.MkfsExt4("foo.img", "my-label", "contents")
+	c.Assert(err, ErrorMatches, "command failed")
+}
+
+func (m *mkfsSuite) TestMkfsVfatHappySimple(c *C) {
+	// no contents, should not fail
+	d := c.MkDir()
+
+	cmd := testutil.MockCommand(c, "mkfs.vfat", "")
+	defer cmd.Restore()
+
+	err := gadget.MkfsVfat("foo.img", "my-label", d)
+	c.Assert(err, IsNil)
+	c.Check(cmd.Calls(), DeepEquals, [][]string{
+		{
+			"mkfs.vfat",
+			"-S", "512",
+			"-s", "1",
+			"-F", "32",
+			"-n", "my-label",
+			"foo.img",
+		},
+	})
+
+	cmd.ForgetCalls()
+
+	// empty label
+	err = gadget.MkfsVfat("foo.img", "", d)
+	c.Assert(err, IsNil)
+	c.Check(cmd.Calls(), DeepEquals, [][]string{
+		{
+			"mkfs.vfat",
+			"-S", "512",
+			"-s", "1",
+			"-F", "32",
+			"foo.img",
+		},
+	})
+}
+
+func (m *mkfsSuite) TestMkfsVfatHappyContents(c *C) {
+	d := c.MkDir()
+	makeSizedFile(c, filepath.Join(d, "foo"), 128, []byte("foo foo foo"))
+	makeSizedFile(c, filepath.Join(d, "bar/bar-content"), 128, []byte("bar bar bar"))
+
+	cmdMkfs := testutil.MockCommand(c, "mkfs.vfat", "")
+	defer cmdMkfs.Restore()
+
+	cmdMcopy := testutil.MockCommand(c, "mcopy", "")
+	defer cmdMcopy.Restore()
+
+	err := gadget.MkfsVfat("foo.img", "my-label", d)
+	c.Assert(err, IsNil)
+	c.Assert(cmdMkfs.Calls(), HasLen, 1)
+
+	c.Assert(cmdMcopy.Calls(), DeepEquals, [][]string{
+		{"mcopy", "-s", "-i", "foo.img", filepath.Join(d, "bar"), filepath.Join(d, "foo"), "::"},
```

Comment:
Having an integration test for this at some point would be cool. But probably not in this PR :)